### PR TITLE
MP-2382. Managed vault

### PR DESCRIPTION
**Pull request by piobab**
Managed vault with TokenFactory support allowing deposit/redeem funds.

The vault is binded to credit manager account (ONLY ONE). Flow to bind Vault:
- instantiate vault contract with credit manager addr
- create new FundManager credit account with vault addr provided. This automatically binds account with vault.

### CODE REVIEW

File: contracts/credit-manager/tests/tests/test_managed_vault.rs [hunk 1]
```diff
@@ -0,0 +1,506 @@
+use anyhow::Result as AnyResult;
```
Comment (piobab) [inline contracts/credit-manager/tests/tests/test_managed_vault.rs:1]: These test cases should live in `mars-vault` test directory. We don't have common multitest setup for Credit Manager in `mars-testing` package for now so we can't easly reuse it (duplicating only produce more code).

I will extract common tests setup in separate PRs and then we will move these test cases to proper place.

File: contracts/vault/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+[package]
+name          = "mars-vault"
```
Comment (piobab) [inline contracts/vault/Cargo.toml:2]: Open for another name. Maybe `mars-managed-vault`?

File: contracts/vault/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        ExecuteMsg::VaultExtension(msg) => match msg {
+            ExtensionExecuteMsg::BindCreditManagerAccount {
+                account_id,
+            } => execute::bind_credit_manager_account(deps, &info, account_id),
+        },
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> ContractResult<Binary> {
```
Comment (piobab) [inline contracts/vault/src/contract.rs:90]: I will add test coverage for the rest of the queries in separate PRs.

File: contracts/vault/src/execute.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        attr("mint_amount", vault_tokens),
+    ]);
+
+    Ok(vault_token
+        .mint(deps, &env, &vault_token_recipient, vault_tokens)?
+        .add_message(deposit_to_cm)
+        .add_event(event))
+}
+
+pub fn redeem(
```
Comment (piobab) [inline contracts/vault/src/execute.rs:96]: Proper Withdraw logic will be provided in separate PR.

File: contracts/credit-manager/tests/tests/helpers/mock_env.rs [hunk 5]
```diff
@@ -810,7 +815,7 @@ impl MockEnvBuilder {
         self.deploy_vaults();
 
         Ok(MockEnv {
-            app: take(&mut self.app),
+            app: self.app,
```
Comment (piobab) [inline contracts/credit-manager/tests/tests/helpers/mock_env.rs:818]: I had to change to ownership (&mut self -> mut self). Custom AppBuilder for multitest can't be used with `take`.

File: contracts/vault/src/execute.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+pub fn bind_credit_manager_account(
+    deps: DepsMut,
+    info: &MessageInfo,
+    account_id: String,
+) -> Result<Response, ContractError> {
+    let credit_manager = CREDIT_MANAGER.load(deps.storage)?;
+    ensure_eq!(info.sender, credit_manager, ContractError::NotCreditManager {});
+
+    // only one binding allowed
+    let cm_acc_id = CREDIT_MANAGER_ACC_ID.may_load(deps.storage)?;
```
Comment (markonmars) [inline contracts/vault/src/execute.rs:22]: Maybe its just me but I found CREDIT_MANAGER_ACC_ID name a little bit confusting at first, it sounds to me like it is the credit managers account id, but in reality it is the vaults account in credit manager? I think maybe `VAULT_ACC_ID` makes more sense to me

File: contracts/vault/src/execute.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    let cm_addr = CREDIT_MANAGER.load(deps.storage)?;
+    let cm_acc_id = CREDIT_MANAGER_ACC_ID.load(deps.storage)?;
+
+    let postions: StdResult<Positions> = deps.querier.query_wasm_smart(
+        cm_addr,
+        &QueryMsg::Positions {
+            account_id: cm_acc_id,
+        },
+    );
+    let postions = postions?;
```
Comment (markonmars) [inline contracts/vault/src/execute.rs:175]: What purpose does the additional line / assignment serve here, is that intentional?

File: contracts/vault/src/execute.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+fn get_amount_to_unlend(
+    deps: Deps,
+    base_token: String,
+    withraw_amount: Uint128,
+) -> Result<Uint128, ContractError> {
+    let cm_addr = CREDIT_MANAGER.load(deps.storage)?;
+    let cm_acc_id = CREDIT_MANAGER_ACC_ID.load(deps.storage)?;
+
+    let postions: StdResult<Positions> = deps.querier.query_wasm_smart(
```
Comment (markonmars) [inline contracts/vault/src/execute.rs:169]: check spelling - positions :)

File: contracts/vault/src/execute.rs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        },
+    );
+    let postions = postions?;
+    let base_token_deposited = postions
+        .deposits
+        .iter()
+        .filter(|d| d.denom == base_token)
+        .map(|d| d.amount)
+        .sum::<Uint128>();
+    let base_token_lend =
```
Comment (markonmars) [inline contracts/vault/src/execute.rs:182]: Does it make sense to place this call below the `if withdraw_amount <= base_token_deposited` statement?

File: contracts/vault/src/execute.rs [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        actions.push(Action::Reclaim(ActionCoin {
+            denom: base_token.clone(),
+            amount: credit_manager::ActionAmount::Exact(amount_to_unlend),
+        }));
+    }
+
+    actions.push(Action::WithdrawToWallet {
+        coin: ActionCoin {
+            denom: base_token,
+            amount: credit_manager::ActionAmount::Exact(tokens_to_withdraw),
```
Comment (markonmars) [inline contracts/vault/src/execute.rs:136]: we use `credit_manager::ActionAmount` a couple of times in this function, the column width and readability is not an issue IMO but maybe still worth importing at the top of the file?

**Review by markonmars — APPROVED** [review]

File: contracts/vault/Cargo.toml (hunk 2 above)
Comment (piobab) [reply contracts/vault/Cargo.toml:2]: Fair enough. Let's leave `mars-vault`.

File: contracts/vault/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Deposit {
+            amount: _, // don't care about amount, use funds data
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/vault/src/contract.rs:74]: Yes. Current `cw-vault-standard` API contains `amount`. Upcoming version of the API won't have amount (see conversation with Apollo on TG or this PR https://github.com/apollodao/cw-vault-standard/blob/dev/no-fixed-base-token/cw-vault-standard/src/msg.rs)

File: contracts/vault/src/execute.rs [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        // bind credit manager account first
+        return Err(ContractError::CreditManagerAccountNotFound {});
+    };
+
+    // unwrap recipient or use caller's address
+    let vault_token_recipient =
+        recipient.map_or(Ok(info.sender.clone()), |r| deps.api.addr_validate(&r))?;
+
+    // load state
+    let base_vault = NtrnBaseVault::default();
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/vault/src/execute.rs:51]: Changed to just `Vault`.

File: contracts/vault/src/execute.rs [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        return Err(ContractError::CreditManagerAccountNotFound {});
+    };
+
+    // unwrap recipient or use caller's address
+    let recipient = recipient.map_or(Ok(info.sender.clone()), |x| deps.api.addr_validate(&x))?;
+
+    // load state
+    let base_vault = NtrnBaseVault::default();
+    let base_token = base_vault.base_token.load(deps.storage)?;
+    let vault_token = base_vault.vault_token.load(deps.storage)?;
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/vault/src/execute.rs:114]: IMO not worth because we have to return back these params.

File: contracts/vault/src/execute.rs (hunk 6 above)
Comment (piobab) [reply contracts/vault/src/execute.rs:22]: @bobthebuidlr wdyt?

File: contracts/vault/src/execute.rs (hunk 7 above)
Comment (piobab) [reply contracts/vault/src/execute.rs:175]: No :) Changed to:
```
let postions: Positions = deps.querier.query_wasm_smart(
        cm_addr,
        &QueryMsg::Positions {
            account_id: cm_acc_id,
        },
    )?;
```

File: contracts/vault/src/execute.rs (hunk 9 above)
Comment (piobab) [reply contracts/vault/src/execute.rs:182]: Updated.

File: contracts/vault/src/execute.rs (hunk 10 above)
Comment (piobab) [reply contracts/vault/src/execute.rs:136]: Imported.

**Comment by piobab** [comment]
@markonmars thx for review! Applied fixes.

File: contracts/vault/src/execute.rs (hunk 6 above)
Comment (bobthebuidlr) [reply contracts/vault/src/execute.rs:22]: Same, it is pretty confusing to me to read it like this. `VAULT_ACC_ID` would be a better fit indeed.

**Review by bobthebuidlr — APPROVED** [review]

File: contracts/vault/src/execute.rs (hunk 6 above)
Comment (piobab) [reply contracts/vault/src/execute.rs]: Updated. Also fixed the rest of the places where this name existed.

File: contracts/vault/src/contract.rs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        deps.api,
+        OwnerInit::SetInitialOwner {
+            owner: info.sender.into(),
+        },
+    )?;
+
+    let credit_manager = deps.api.addr_validate(&msg.credit_manager)?;
+    CREDIT_MANAGER.save(deps.storage, &credit_manager.to_string())?;
+
+    if let Some(tit) = msg.title {
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/vault/src/contract.rs]: I was hungry ;) Fixed
